src/rgw/rgw_acl_s3.cc

```diff
@@ -433,117 +454,68 @@ static const s3_acl_header acl_header_perms[] = {
   {0, NULL}
 };
 
-/*
-  can only be called on object that was parsed
- */
-int RGWAccessControlPolicy_S3::rebuild(const DoutPrefixProvider *dpp,
-				       rgw::sal::Driver* driver, ACLOwner *owner,
-				       RGWAccessControlPolicy& dest, std::string &err_msg)
+static int resolve_grant(const DoutPrefixProvider* dpp, optional_yield y,
+                         rgw::sal::Driver* driver, ACLGrant_S3& xml_grant,
+                         ACLGrant& grant, std::string& err_msg)
 {
-  if (!owner || owner->id.empty()) {
-    return -EINVAL;
-  }
+  const uint32_t perm = xml_grant.permission->flags;
 
-  ACLOwner *requested_owner = static_cast<ACLOwner_S3 *>(find_first("Owner"));
-  if (requested_owner && requested_owner->id != owner->id) {
-    return -EPERM;
-  }
+  std::unique_ptr<rgw::sal::User> user;
+  switch (xml_grant.type.get_type()) {
+  case ACL_TYPE_EMAIL_USER:
+    if (xml_grant.email.empty()) {
+      return -EINVAL;
+    }
+    if (driver->get_user_by_email(dpp, xml_grant.email, y, &user) < 0) {
```

Comment:
why was this using the `null_yield` before?

Reply:
`RGWAccessControlPolicy_S3::rebuild()` just didn't have an `optional_yield` argument. i added one to the new `parse_policy()` function in this commit